File: 9_AddLL/solution.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def addTwoLists(first, second):
    dummy = Node(0)
    curr = dummy
    carry = 0
    while first or second or carry:
        total = carry
        if first:
            total += first.data
            first = first.next
        if second:
            total += second.data
            second = second.next
        carry = total // 10
        curr.next = Node(total % 10)
        curr = curr.next
    return dummy.next


def build_list():
    head = None
    tail = None
    n = int(input("Enter number of elements: "))
    for _ in range(n):
        val = int(input("Enter value: "))
        new_node = Node(val)
        if not head:
            head = tail = new_node
        else:
            tail.next = new_node
            tail = new_node
    return head

def main():
    head = build_list()

    printList(head)

if __name__ == "__main__":
    main()

File: 9_AddLL/solution.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* addTwoLists(struct Node* first, struct Node* second) {
    struct Node* res = NULL;
    struct Node** node = &res;
    int carry = 0;
    while (first || second || carry) {
        int sum = carry;
        if (first) { sum += first->data; first = first->next; }
        if (second) { sum += second->data; second = second->next; }
        carry = sum / 10;
        *node = (struct Node*) malloc(sizeof(struct Node));
        (*node)->data = sum % 10;
        node = &((*node)->next);
    }
    *node = NULL;
    return res;
}


int main() {
    struct Node* head = NULL;
    int n, val, pos;
    printf("Enter number of elements: ");
    scanf("%d", &n);
    for (int i = 0; i < n; i++) {
        printf("Enter value: ");
        scanf("%d", &val);
        head = insertEnd(head, val);
    }

    printList(head);
    return 0;
}